camel/toolkits/whatsapp_toolkit.py

```diff
@@ -0,0 +1,185 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+
+import os
+import time
+from typing import Any, Dict, List, Union
+
+import requests
+
+from camel.toolkits import OpenAIFunction
+from camel.toolkits.base import BaseToolkit
+
+
+class WhatsAppToolkit(BaseToolkit):
+    """A class representing a toolkit for WhatsApp operations.
+
+    This toolkit provides methods to interact with the WhatsApp Business API,
+    allowing users to send messages, retrieve message templates, and get
+    business profile information.
+
+    Attributes:
+        retries (int): Number of retries for API requests in case of failure.
+        delay (int): Delay between retries in seconds.
+        base_url (str): Base URL for the WhatsApp Business API.
+        version (str): API version.
+    """
+
+    def __init__(self, retries: int = 3, delay: int = 1):
+        """Initializes the WhatsAppToolkit with the specified number of retries
+        and delay.
+
+        Args:
+            retries (int): Number of times to retry the request in case of
+                failure. Defaults to 3.
+            delay (int): Time in seconds to wait between retries.
+            Defaults to 1.
+        """
+        self.retries = retries
+        self.delay = delay
+        self.base_url = "https://graph.facebook.com"
+        self.version = "v17.0"
+
+        self.access_token = os.environ.get("WHATSAPP_ACCESS_TOKEN", "")
+        self.phone_number_id = os.environ.get("WHATSAPP_PHONE_NUMBER_ID", "")
+
+        if not all([self.access_token, self.phone_number_id]):
+            raise ValueError(
+                "WhatsApp API credentials are not set. "
+                "Please set the WHATSAPP_ACCESS_TOKEN and "
+                "WHATSAPP_PHONE_NUMBER_ID environment variables."
+            )
+
+    def _retry_request(self, func, *args, **kwargs):
+        """Retries a function in case of any errors.
+
+        Args:
+            func (callable): The function to be retried.
+            *args: Arguments to pass to the function.
+            **kwargs: Keyword arguments to pass to the function.
+
+        Returns:
+            Any: The result of the function call if successful.
+
+        Raises:
+            Exception: If all retry attempts fail.
+        """
+        for attempt in range(self.retries):
+            try:
+                return func(*args, **kwargs)
+            except Exception as e:
+                print(f"Attempt {attempt + 1}/{self.retries} failed: {e}")
+                if attempt < self.retries - 1:
+                    time.sleep(self.delay)
+                else:
+                    raise
```

Comment:
since this method can be generally used, we can move this method under `camel/utils/commons.py`

Reply:
Moved this to `camel/utils/commons.py`. Please review again. Thanks.